nomad/state/state_store_host_volumes.go

```diff
@@ -190,3 +190,35 @@ func (s *StateStore) hostVolumesIter(ws memdb.WatchSet, index string, sort SortO
 	ws.Add(iter.WatchCh())
 	return iter, nil
 }
+
+// upsertHostVolumeForNode sets newly fingerprinted host volumes to ready state
+func upsertHostVolumeForNode(txn *txn, node *structs.Node, index uint64) error {
+	if len(node.HostVolumes) == 0 {
+		return nil
+	}
+	iter, err := txn.Get(TableHostVolumes, indexNodeID, node.ID)
+	if err != nil {
+		return err
+	}
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			return nil
+		}
+		vol := raw.(*structs.HostVolume)
+		switch vol.State {
+		case structs.HostVolumeStateUnknown, structs.HostVolumeStatePending:
+			if _, ok := node.HostVolumes[vol.Name]; ok {
+				vol = vol.Copy()
+				vol.State = structs.HostVolumeStateReady
+				vol.ModifyIndex = index
+				err = txn.Insert(TableHostVolumes, vol)
+				if err != nil {
+					return fmt.Errorf("host volume insert: %w", err)
+				}
+			}
+		default:
+			// don't touch ready or soft-deleted volumes
```

Comment:
what's meant by "soft-deleted" here?  deleted on the client without a respective ->server->rpc->state call?

Reply:
There's a `HostVolumeStateDeleted` in our state enum. That gives us the option to delete the volume from the client via the RPC but leave the record of it on the server until GC'd, with the reasoning being that operators might care about being able to track that sort of thing. The RPC doesn't currently implement that, because I haven't quite reasoned though the consequences of GC'ing them properly.